compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -184,3 +174,155 @@ prettyFileDiagnostics (filePath, diags) =
 getDiagnosticsFromStore :: StoreItem -> [Diagnostic]
 getDiagnosticsFromStore (StoreItem _ diags) =
     toList =<< Map.elems diags
+
+-- | This represents every diagnostic in a LSP project, the stage type variable is
+--   the type of the compiler stages, in this project that is always the Key data
+--   type found in Development.IDE.State.Shake
+newtype ProjectDiagnostics stage = ProjectDiagnostics {getStore :: DiagnosticStore}
+    deriving Show
+
+prettyDiagnostics :: ProjectDiagnostics stage -> Doc SyntaxClass
+prettyDiagnostics ds =
+    slabel_ "Compiler errors in" $ vcat $ concatMap fileErrors storeContents where
+
+    fileErrors :: (FilePath, [(T.Text, [LSP.Diagnostic])]) -> [Doc SyntaxClass]
+    fileErrors (filePath, stages) =
+        [ slabel_ "File:" $ pretty filePath
+        , slabel_ "Errors:" $ vcat $ map prettyStage stages
+        ]
+
+    prettyStage :: (T.Text, [LSP.Diagnostic]) -> Doc SyntaxClass
+    prettyStage (stage,diags) =
+        slabel_ ("Stage: "<>T.unpack stage) $ vcat $ concatMap diagSC diags
+
+    storeContents ::
+        [(FilePath, [(T.Text, [LSP.Diagnostic])])]
+        -- ^ Source File, Stage Source, Diags
+    storeContents =
+        map (\(uri, StoreItem _ si) ->
+                 (fromMaybe noFilePath $ uriToFilePath uri, getDiags si))
+            $ Map.assocs
+            $ getStore
+            $ removeEmptyStages ds
+
+    getDiags :: DiagnosticsBySource -> [(T.Text, [LSP.Diagnostic])]
+    getDiags = map (\(ds, diag) -> (fromMaybe (T.pack noFilePath) ds, toList diag)) . Map.assocs
+
+emptyDiagnostics :: ProjectDiagnostics stage
+emptyDiagnostics = ProjectDiagnostics mempty
+
+-- | Sets the diagnostics for a file and compilation step
+--   if you want to clear the diagnostics call this with an empty list
+setStageDiagnostics ::
+  Show stage =>
```

Comment:
This is odd formatting. Everyone else puts the `->` at the beginning of the line.

Reply:
I'm quite a fan, it gives you alignment without white space padding, while making it clear that this is a constraint. That being said consistency is probably more important than correctness